File: scripts/cpd_report.py
#!/usr/bin/env python3
"""CPD (Copy-Paste Detector) XML report formatter.

Reads PMD CPD XML output from stdin, prints a human-readable summary, and
exits with code 1 when --strict is given and any duplication clusters are found.

Usage:
  pmd cpd --language rust --minimum-tokens 42 --dir crates --format xml 2>/dev/null \
    | python3 scripts/cpd_report.py [--strict] [--top N]

Exit codes:
  0  No clusters found, or --strict not given.
  1  --strict given and at least one cluster found.
  2  PMD/XML parse error (empty or malformed input).
"""

import sys
import xml.etree.ElementTree as ET
import argparse


def short_path(path: str) -> str:
    for marker in ("crates/", "src/"):
        idx = path.find(marker)
        if idx >= 0:
            return path[idx:]
    return path


def main() -> int:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--strict", action="store_true",
                        help="Exit 1 if any clusters are found.")
    parser.add_argument("--top", type=int, default=20,
                        help="Show top N clusters by line count (default: 20).")
    args = parser.parse_args()

    raw = sys.stdin.read()
    if not raw.strip():
        print("cpd-report: no XML received (pmd produced no output).", file=sys.stderr)
        return 2

    try:
        root = ET.fromstring(raw)
    except ET.ParseError as exc:
        print(f"cpd-report: XML parse error: {exc}", file=sys.stderr)
        return 2

    dups = root.findall("duplication")
    total_lines = sum(int(d.get("lines", 0)) for d in dups)

    print(f"CPD audit  --minimum-tokens 42  Rust")
    print(f"Clusters : {len(dups)}")
    print(f"Dup lines: {total_lines}")

    if not dups:
        print("No duplication found.")
        return 0

    # Sort by line count descending, show top N.
    dups_sorted = sorted(dups, key=lambda d: int(d.get("lines", 0)), reverse=True)
    show = dups_sorted[: args.top]

    print()
    print(f"{'#':<4}  {'Lines':>6}  {'Tokens':>7}  Locations")
    print("-" * 72)
    for i, dup in enumerate(show, 1):
        lines = int(dup.get("lines", 0))
        tokens = int(dup.get("tokens", 0))
        locs = [
            f"{short_path(f.get('path','?'))}:{f.get('line','?')}"
            for f in dup.findall("file")
        ]
        loc_str = "  /  ".join(locs)
        print(f"{i:<4}  {lines:>6}  {tokens:>7}  {loc_str}")
        frag = (dup.findtext("codefragment") or "").strip()
        if frag:
            preview = frag.replace("\n", " ")[:90]
            print(f"       {preview!r}")

    if len(dups) > args.top:
        print(f"  ... {len(dups) - args.top} more clusters not shown.")

    if args.strict:
        print(f"\ncpd-audit-strict: FAIL -- {len(dups)} cluster(s) found.", file=sys.stderr)
        return 1

    return 0


if __name__ == "__main__":
    sys.exit(main())

File: crates/gororoba_cli_data/src/bin/cpd_report.rs
use anyhow::Result;
use clap::Parser;
use quick_xml::{events::Event, reader::Reader};
use std::io::{self, Read};

#[derive(Parser, Debug)]
#[command(
    name = "cpd-report",
    about = "PMD CPD XML report formatter (Rust replacement for cpd_report.py)"
)]
struct Cli {
    #[arg(long)]
    strict: bool,
    #[arg(long, default_value_t = 20)]
    top: usize,
}

#[derive(Debug, Default)]
struct Duplication {
    lines: usize,
    tokens: usize,
    files: Vec<CpdFile>,
    codefragment: String,
}

#[derive(Debug, Default)]
struct CpdFile {
    path: String,
    line: usize,
}

fn short_path(path: &str) -> &str {
    for marker in &["crates/", "src/"] {
        if let Some(idx) = path.find(marker) {
            return &path[idx..];
        }
    }
    path
}

fn main() -> Result<()> {
    let cli = Cli::parse();

    let mut input = String::new();
    io::stdin().read_to_string(&mut input)?;

    if input.trim().is_empty() {
        eprintln!("cpd-report: no XML received (pmd produced no output).");
        std::process::exit(2);
    }

    let mut reader = Reader::from_str(&input);
    let mut dups = Vec::new();
    let mut current_dup: Option<Duplication> = None;
    let mut buf = Vec::new();

    loop {
        match reader.read_event_into(&mut buf) {
            Ok(Event::Start(e)) => match e.name().as_ref() {
                b"duplication" => {
                    let mut dup = Duplication::default();
                    for attr in e.attributes().flatten() {
                        match attr.key.as_ref() {
                            b"lines" => dup.lines = std::str::from_utf8(&attr.value)?.parse()?,
                            b"tokens" => dup.tokens = std::str::from_utf8(&attr.value)?.parse()?,
                            _ => {}
                        }
                    }
                    current_dup = Some(dup);
                }
                b"file" => {
                    if let Some(ref mut dup) = current_dup {
                        let mut file = CpdFile::default();
                        for attr in e.attributes().flatten() {
                            match attr.key.as_ref() {
                                b"path" => {
                                    file.path = std::str::from_utf8(&attr.value)?.to_string()
                                }
                                b"line" => file.line = std::str::from_utf8(&attr.value)?.parse()?,
                                _ => {}
                            }
                        }
                        dup.files.push(file);
                    }
                }
                b"codefragment" => {
                    if let Some(ref mut dup) = current_dup {
                        dup.codefragment = reader.read_text(e.name())?.to_string();
                    }
                }
                _ => {}
            },
            Ok(Event::End(e)) if e.name().as_ref() == b"duplication" => {
                if let Some(dup) = current_dup.take() {
                    dups.push(dup);
                }
            }
            Ok(Event::Eof) => break,
            Err(e) => {
                eprintln!("cpd-report: XML parse error: {}", e);
                std::process::exit(2);
            }
            _ => {}
        }
        buf.clear();
    }

    let total_lines: usize = dups.iter().map(|d| d.lines).sum();

    println!("CPD audit  --minimum-tokens 42  Rust");
    println!("Clusters : {}", dups.len());
    println!("Dup lines: {}", total_lines);

    if dups.is_empty() {
        println!("No duplication found.");
        return Ok(());
    }

    dups.sort_by_key(|d| std::cmp::Reverse(d.lines));
    let show = &dups[..std::cmp::min(dups.len(), cli.top)];

    println!();
    println!("{: <4}  {: >6}  {: >7}  Locations", "#", "Lines", "Tokens");
    println!("{}", "-".repeat(72));

    for (i, dup) in show.iter().enumerate() {
        let locs: Vec<String> = dup
            .files
            .iter()
            .map(|f| format!("{}:{}", short_path(&f.path), f.line))
            .collect();
        let loc_str = locs.join("  /  ");
        println!(
            "{: <4}  {: >6}  {: >7}  {}",
            i + 1,
            dup.lines,
            dup.tokens,
            loc_str
        );
        let preview = dup.codefragment.replace('\n', " ");
        let preview = if preview.len() > 90 {
            format!("{}...", &preview[..87])
        } else {
            preview
        };
        println!("       {:?}", preview);
    }

    if dups.len() > cli.top {
        println!("  ... {} more clusters not shown.", dups.len() - cli.top);
    }

    if cli.strict && !dups.is_empty() {
        eprintln!(
            "\ncpd-audit-strict: FAIL -- {} cluster(s) found.",
            dups.len()
        );
        std::process::exit(1);
    }

    Ok(())
}
